### PR TITLE
Move some helper functions from api/v1 to component-helpers

### PR DESCRIPTION
**Pull request by AxeZhan**
#### What type of PR is this?
/kind cleanup
/kind feature

#### What this PR does / why we need it:
We currently have some functions which are used by many components,
for example:
https://github.com/kubernetes/kubernetes/blob/ae02f87bb47cd3e10e702ffe19225ff2dba73578/pkg/api/v1/resource/helpers.go#L50

It would be useful to have them in a library package.
#### Which issue(s) this PR fixes:

Part of #124537

#### Special notes for your reviewer:
First commit moves the file.
Second commit only do the rename.
#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/component-helpers/pod/util.go
Comment (neolit123) [inline staging/src/k8s.io/component-helpers/pod/util.go:2]: it seems files with utils in component-helpers are called "helpers.go"

File: pkg/controller/controller_utils.go
Comment (alculquicondor) [inline pkg/controller/controller_utils.go:47]: Each new library has to go inside of the folders owned by a SIG.

File: staging/src/k8s.io/component-helpers/pod/util.go
Comment (alculquicondor) [inline staging/src/k8s.io/component-helpers/pod/util.go:1]: I would prefer we just target the functions that we need for resource calculations, and nothing else.

Comment (AxeZhan) [reply staging/src/k8s.io/component-helpers/pod/util.go:1]: I removed `PodRequests`, `PodLimits` and leave the rest as it is.

File: pkg/controller/controller_utils.go
Comment (AxeZhan) [reply pkg/controller/controller_utils.go]: Added a `OWNERS` file, not sure if it's sig-node that should own this folder.

File: staging/src/k8s.io/component-helpers/pod/util.go
Comment (AxeZhan) [reply staging/src/k8s.io/component-helpers/pod/util.go]: Renamed file to `helpers.go`.

File: staging/src/k8s.io/component-helpers/resource/OWNERS
Comment (SergeyKanzhelev) [reply staging/src/k8s.io/component-helpers/resource/OWNERS:1]: sig node is a good here. InPlace and Sidecar are both driven by SIG Node and making changes to this.  But we can share with the sig scheduling.

Comment (SergeyKanzhelev) [reply staging/src/k8s.io/component-helpers/resource/OWNERS:1]: I am thinking if this must be considered a public API and api-approvers needs to be notified for changes here

File: staging/src/k8s.io/component-helpers/resource/helpers.go
Comment (SergeyKanzhelev) [reply staging/src/k8s.io/component-helpers/resource/helpers.go]: +1

File: staging/src/k8s.io/component-helpers/resource/OWNERS
Comment (SergeyKanzhelev) [reply staging/src/k8s.io/component-helpers/resource/OWNERS:1]: @kubernetes/api-reviewers should we add api review label for changes in these files? Or need to add it manually?

(we added the note The computation is part of the API and must be reviewed as an API change. to the function, but not sure if this is enough)

Comment (liggitt) [reply staging/src/k8s.io/component-helpers/resource/OWNERS:1]: Yeah, not just notified, but approval constrained to API review. We've avoided relocating API validation functions out of API packages for that reason in the past before… it's too easy to accidentally tighten or loosen validation in ways that cause issues without realizing it when the functions are in packages with more broad approval.

That's not to say that relocation can't happen here, we just need to make sure changes are done as carefully as needed

File: staging/src/k8s.io/component-helpers/resource/OWNERS
Comment (alculquicondor) [inline staging/src/k8s.io/component-helpers/resource/OWNERS:4]: IIUC from the discussion in the other thread with @SergeyKanzhelev and @liggitt,
approvers should be `api-approvers` only.

Reviewers can be `sig-node-reviewers`, `sig-scheduling` (this is the reviewers list... yes it should be renamed).

Labels: `sig/node` and `sig/scheduling`

Comment (AxeZhan) [reply staging/src/k8s.io/component-helpers/resource/OWNERS]: Updated

File: pkg/kubelet/container/helpers.go
Comment (alculquicondor) [inline pkg/kubelet/container/helpers.go:33]: For new imports, only use an alias if there is a conflict.

Same for other files.

File: staging/src/k8s.io/component-helpers/resource/helpers.go
Comment (alculquicondor) [inline staging/src/k8s.io/component-helpers/resource/helpers.go:55]: Question for @liggitt: would it be fine to also incorporate the logic of "requests = limits, when not defined".

https://github.com/kubernetes/kubernetes/blob/b041e54e70fc906eb753b6acc6967530cd3fb720/pkg/apis/core/v1/defaults.go#L158

File: pkg/kubelet/container/helpers.go
Comment (liggitt) [inline pkg/kubelet/container/helpers.go:313]: why this change? previously this explicitly excluded any container types whose corresponding feature was not enabled

File: staging/src/k8s.io/component-helpers/resource/helpers.go
Comment (liggitt) [reply staging/src/k8s.io/component-helpers/resource/helpers.go:55]: two thoughts:
1. don't mix relocation and modification into the same PR
2. why would we need to duplicate API defaulting logic here? admission / controllers / clients reading from the API will already see pods with the defaults applied

File: staging/src/k8s.io/component-helpers/resource/OWNERS
Comment (liggitt) [reply staging/src/k8s.io/component-helpers/resource/OWNERS:1]: is there a diff of the content in the old location and new location to make sure no functionality is changing in the move for `pkg/api/v1/resource/helpers.go` and `pkg/api/v1/resource/helpers_test.go`?

File: staging/src/k8s.io/component-helpers/resource/OWNERS
Comment (liggitt) [inline staging/src/k8s.io/component-helpers/resource/OWNERS:3]: also needs this to effectively bound to API review

```yaml
# Disable inheritance as this is an api owners file
options:
  no_parent_owners: true
```

File: staging/src/k8s.io/component-helpers/resource/OWNERS
Comment (liggitt) [inline staging/src/k8s.io/component-helpers/resource/OWNERS:10]: add
```
  - kind/api-change
```

File: pkg/api/pod/util.go
Comment (liggitt) [inline pkg/api/pod/util.go:51]: why was this dropped? this gives callers a way to express "I want all the container types whose features are enabled"

Comment (AxeZhan) [reply pkg/api/pod/util.go:51]: We thought this was a legacy function: https://github.com/kubernetes/kubernetes/pull/124609#discussion_r1584798745
Maybe we can add it back when there is a new feature-related container type?

Comment (liggitt) [reply pkg/api/pod/util.go:51]: no, that will require revisiting all uses of `AllContainers` and trying to decide again whether they care about honoring feature-disablement

File: staging/src/k8s.io/component-helpers/resource/OWNERS
Comment (AxeZhan) [reply staging/src/k8s.io/component-helpers/resource/OWNERS:1]: here're the diff links:
[helpers.go](https://www.diffchecker.com/J47igkHy/)
[helpers_test.go](https://www.diffchecker.com/yINhucNZ/)

File: staging/src/k8s.io/component-helpers/resource/helpers.go
Comment (alculquicondor) [reply staging/src/k8s.io/component-helpers/resource/helpers.go:67]: 1. I agree.
2. Because users of this helper might be using a Pod template that didn't pass through the apiserver yet. For example, when using an external CRD such as a kubeflow TFJob.

File: pkg/api/pod/util.go
Comment (alculquicondor) [reply pkg/api/pod/util.go]: why not use the `AllContainers` constant directly?

File: staging/src/k8s.io/component-helpers/resource/helpers.go
Comment (liggitt) [reply staging/src/k8s.io/component-helpers/resource/helpers.go:67]: but... the input to this function is Pod, not PodTemplate

File: pkg/api/pod/util.go
Comment (liggitt) [reply pkg/api/pod/util.go]: AllFeatureEnabledContainers() and AllContainers *happen* to be the same right now, but all callers of AllFeatureEnabledContainers() are intending to ignore any container types disabled by feature gates. Collapsing them to use AllContainers directly loses that information and means any future feature-gates have to try to recover it.

File: staging/src/k8s.io/component-helpers/resource/OWNERS
Comment (AxeZhan) [reply staging/src/k8s.io/component-helpers/resource/OWNERS:1]: @liggitt Can you review again? Diff files seem fine. 
But notice that I move `ContainerType` from `pkg/api/pod/util.go` to `component-helpers/resource/helpers.go` too. Becuase `PodRequests` function uses `ContainerType`, and import pkg `v1` in `staging` is not allowed.

File: staging/src/k8s.io/component-helpers/resource/helpers.go
Comment (alculquicondor) [reply staging/src/k8s.io/component-helpers/resource/helpers.go:67]: Yeah, but one can easily convert one to the other in order to get a resource calculation.

**Comment by krasaee** [comment]
@AxeZhan @alculquicondor  -- apologize for any ignorant comments/questions in advance -- yet: 

Should this helper definition [e.g. podRequests(..)] even be in mod kubernetes k8s.io/kubernetes/pkg/api/v1/resource? 

I thought referencing kubernetes mod (ex: we use k8s.io/kubernetes/pkg/api/v1/resource) is not recommended - the kubernetes repo states not to use it as a module/library, rather reference the individual components separately. This makes sense, to me atleast, but also resolves a bunch of v0.0.0 dependency issues if I reference kubernetes mod directly.

So on one hand we have a helper method that is referenced (by internal code we use - which can be changed) and on the other hand we should not be using it. 

image

**Comment by alculquicondor** [comment]
That's why this PR is moving it out of k8s.io/kubernetes

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve

**Comment by alculquicondor** [comment]
@AxeZhan do you still have time to work on this?

**Comment by AxeZhan** [comment]
> @AxeZhan do you still have time to work on this?

Yes, I'll do a rebase recently.

File: staging/src/k8s.io/component-helpers/resource/OWNERS
Comment (alculquicondor) [reply staging/src/k8s.io/component-helpers/resource/OWNERS:1]: Can you update the diff links? Since the last ones are from May.

File: staging/src/k8s.io/component-helpers/resource/helpers.go
Comment (alculquicondor) [inline staging/src/k8s.io/component-helpers/resource/helpers.go:52]: Is there anything using `ExcludeOverhead=true`?

**Comment by AxeZhan** [comment]
Diff of helpers.go: https://www.diffchecker.com/0C7TdlbC/

File: staging/src/k8s.io/component-helpers/resource/helpers.go
Comment (AxeZhan) [reply staging/src/k8s.io/component-helpers/resource/helpers.go:52]: I think so.
https://github.com/kubernetes/kubernetes/blob/f248c24456254d6e0668a424a31258e6d9e9c185/pkg/kubelet/cm/devicemanager/topology_hints.go#L237

Comment (alculquicondor) [reply staging/src/k8s.io/component-helpers/resource/helpers.go:52]: @ffromani suggests that we keep this https://github.com/kubernetes/kubernetes/commit/4096c9209cbf20c51d184e83ab6ffa3853bd2ee6#r146822193

File: staging/src/k8s.io/component-helpers/resource/helpers.go
Comment (alculquicondor) [reply staging/src/k8s.io/component-helpers/resource/helpers.go:76]: If anything, let's leave this for a follow up.

File: staging/src/k8s.io/component-helpers/resource/helpers_test.go
Comment (alculquicondor) [inline staging/src/k8s.io/component-helpers/resource/helpers_test.go:1132]: nit: would `cmp.Diff` have the same effect? I believe it would.

**Comment by alculquicondor** [comment]
@SergeyKanzhelev @liggitt PTAL

**Comment by SergeyKanzhelev** [comment]
> @SergeyKanzhelev @liggitt PTAL

@AxeZhan thank you for driving this. Can you please post diff links to the latest iterations to see there is no difference in behavior in this move?

**Comment by AxeZhan** [comment]
@SergeyKanzhelev diff link is posted at https://github.com/kubernetes/kubernetes/pull/124609#issuecomment-2333975327

I'm not sure if this can be done on github itself, so I just googled a website for it😢

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

looks good for me. Code was moved unmodified so it is mostly the packaging problem. And making it reusable sounds great

**Comment by SergeyKanzhelev** [comment]
/priority important-soon

with the https://github.com/kubernetes/enhancements/issues/2837 coming soon, merging this will be very helpful to avoid copy/paste mistakes

**Comment by alculquicondor** [comment]
/hold

As we found a bug related to sidecar containers #127878

**Comment by liggitt** [comment]
moving to API review backlog until prereq is complete

**Comment by AxeZhan** [comment]
/hold cancel
https://github.com/kubernetes/kubernetes/pull/127878 merged and rebased.

**Comment by AxeZhan** [comment]
Newly diff: https://www.diffchecker.com/aievbqFe/
There's no conflict in the first commit, so it should stays the same

File: staging/src/k8s.io/component-helpers/resource/helpers.go
Comment (pacoxu) [inline staging/src/k8s.io/component-helpers/resource/helpers.go:30]: Should we mention that the current init containers includes sidecar containers in the comment? Not have to. The pod limit\request calculation session has already include the KEP link below. https://github.com/kubernetes/kubernetes/pull/124609/files#diff-22767ad86bae6a15ac904d98655b9f1eff79bf51470b419c85c1d09d41cdb011R108

Comment (alculquicondor) [reply staging/src/k8s.io/component-helpers/resource/helpers.go:30]: This relates to the question I left on the other PR https://github.com/kubernetes/kubernetes/pull/127878#issuecomment-2405521951

Should we be worried about other users of ContainerFn. Maybe it's worth adding SidecarContainers before putting this method in the helpers package?

File: staging/src/k8s.io/component-helpers/resource/helpers.go
Comment (ndixita) [reply staging/src/k8s.io/component-helpers/resource/helpers.go:76]: Another follow up question: Is there a reason whey PodRequests has the entire Pod object as the parameter instead of PodSpec and PodStatus explicitly? I wanted to modify the Validation logic of PodTemplate to use PodRequests to validate pod-level resources validation criteria and hence wondering...

File: staging/src/k8s.io/component-helpers/resource/helpers.go
Comment (AxeZhan) [reply staging/src/k8s.io/component-helpers/resource/helpers.go:30]: I think it's better add `SidecarContainers` when some developer actually needs to use it?

Comment (gjkim42) [reply staging/src/k8s.io/component-helpers/resource/helpers.go:30]: https://github.com/kubernetes/kubernetes/pull/116429#discussion_r1235351189

@smarterclayton and I had a similar discussion.

I am worried about explicitly using SidecarContainer naming (other than the feature name), as the "sidecar" is quite a general term.

I think the point of introducing the "Always" containerRestartPolicy was not to create one concrete "SidecarContainer" type. It just updated the restart behavior of the existing init containers so that kubernetes can have native way to support sidecar **pattern**.

**Comment by alculquicondor** [comment]
@liggitt could you assign an api reviewer?

File: staging/src/k8s.io/component-helpers/resource/helpers.go
Comment (liggitt) [reply staging/src/k8s.io/component-helpers/resource/helpers.go:30]: I wouldn't change this here / now in this PR

sidecar containers *are* init containers, just of a special type... if we want to clarify that later in a comment, that could be fine

Comment (liggitt) [reply staging/src/k8s.io/component-helpers/resource/helpers.go:30]: > Should we be worried about other users of ContainerFn? Maybe it's worth adding SidecarContainers before putting this method in the helpers package?

I think this PR should be a *strictly* mechanical move, not change semantics

File: pkg/api/v1/pod/util.go
Comment (liggitt) [inline pkg/api/v1/pod/util.go:54]: making the container iteration functions use constants from the resource helper package is really weird... I don't see the value of sharing types here. I think we should restore the ContainerType / Containers/InitContainers/EphemeralContainers / AllContainers / AllFeatureEnabledContainers bits to this package, and narrowly define the two constants we need in the resource helper package (Containers/InitContainers)


that would also drop out ~all the unrelated package import changes in this PR for things which aren't actually doing anything with resource calculation

**Comment by liggitt** [comment]
(the diff on the relocated resource helper functions looks good... I verified it's a zero delta)

File: pkg/api/v1/pod/util.go
Comment (AxeZhan) [reply pkg/api/v1/pod/util.go]: Updated :)

**Comment by AxeZhan** [comment]
Diff link:
https://www.diffchecker.com/oho4jC4l/

**Pull request merged**
